### ISSUE TITLE
Some Unit tests when run from the device are failing.

### ISSUE THREAD

**Issue by ashvindersingh** [couchbase/couchbase-lite-ios#874]
Some of the Unit tests when run on device are failing due to unable/trying to connect with syncgateway url 127.0.0.1 ipaddress. Instead they should be sending data to provisioned server in xcscheme file.

**Comment by snej** [comment]
I don't think the tests have any hardcoded URLs; they should be honoring the `CBL_TEST_SERVER` and `CBL_SSL_TEST_SERVER` environment variables. (Except of course for P2P and listener tests that hardcode `localhost` because they run both client and listener in-process.)

Have you verified that the scheme is setting those environment variables?

**Comment by snej** [comment]
In the commit above it looks like you enabled those variables but left them with their default values that point to `localhost`. For this to fix the on-device testing, you have to change the host of those URLs to point to the machine that's running the Sync Gateway.

**Comment by ashvindersingh** [comment]
Hi @snej, I deliberately did it so that Unit tests can pass.  I'm replacing the 'localhost' with sync gateway URL in bots configuration. Now we have 7 tests which are failing with the device. The remaining replication tests are passing. I have discussed with @pasin. He is looking into it.

**Comment by pasin** [comment]
There are 3 tests failed due to an invalid SSL certificate. To this with devices, the self-signed certificate needs to be regenerated with the name of the Sync-Gateway IP Address.

```
$openssl genrsa -out privkey.pem 2048
$openssl req -new -x509 -key privkey.pem -out cert.pem -days 1095 
<Use Sync-Gateway IP Address as the name>
$openssl x509 -inform PEM -in cert.pem -outform DER -out SelfSigned.cer
```
Replace TestData/SelfSigned.cer, Server/cert.pem, and Server/privkey.pem with the re-generated ones.

**Comment by pasin** [comment]
For the other failures:
- Database_Tests.test23_ReplaceOldVersionDatabase (#875)
- Misc_Tests.test_AnonymousIdentity (Couldn't reproduce the failure)
- DatabaseInternal_Tests.test25_FileProtection (Pending investigation)

I have also found that IncrementalStore_test.test_DefaultConflictHandler failed sometime. I will have to investigate this.

**Comment by snej** [comment]
>There are 3 tests failed due to an invalid SSL certificate.

Yup, I've been pointing this out for as long as the bot has been running (and failing) those tests.

**Comment by pasin** [comment]
For DatabaseInternal_Tests.test25_FileProtection, the cause of the failure was that the SQLite -shm file has NSFileProtectionNone.
 
@snej I did google and have found your 2012 post (http://lists.apple.com/archives/cocoa-dev/2012/Aug/msg00512.html), is this a known issue?

**Comment by snej** [comment]
We set the file protection of the database's directory, and that's supposed to apply to all files created in the directory. Even if it doesn't, I would think the temp SQLite files would at least get the app's default protection level, which is something greater than none.

**Comment by pasin** [comment]
I found a blog [post](http://movq.us/2014/10/10/core-data-and-data-protection/) that encountered a similar issue but with Core Data SQLite. Core Data's NSPersistentStoreCoordinator has an option to set the file protection level (e.g. @{NSPersistentStoreFileProtectionKey:NSFileProtectionComplete} ) that will open the database with the corresponding flag (e.g SQLITE_OPEN_FILEPROTECTION_COMPLETE).

**Comment by pasin** [comment]
I have tested this by setting different SQLite open flags including:
SQLITE_OPEN_FILEPROTECTION_COMPLETE, SQLITE_OPEN_FILEPROTECTION_COMPLETEUNLESSOPEN,
SQLITE_OPEN_FILEPROTECTION_COMPLETEUNTILFIRSTUSERAUTHENTICATION.

SQLite files (.sqlite, -wal, -shm) created by the database have same file protection level as the given flag EXCEPT the `-shm` file. The `-shm` file always has `NSFileProtectionCompleteUntilFirstUserAuthentication` level. However, without specifying the file protection flag, the `-shm` file has NSFileProtectionNone level while the other files (.sqlite and -wal) have the same protection level as its parent folder. (This result is inline with the blog post that I posted in my previous comment).

I couldn't find the information describing why the protection level of the -shm file is always at `NSFileProtectionCompleteUntilFirstUserAuthentication` level. But from the [SQLite document](https://www.sqlite.org/tempfiles.html) about -shm file,
> The shared-memory file contains no persistent content. The only purpose of the shared-memory file is to provide a block of shared memory for use by multiple processes all accessing the same database in WAL mode.

it seems to be fine because without accessing to the -wal file, -shm file is not useful, and more importantly it contains no persistent content.

I think we should set the correct file protection level when opening the SQLite database anyway.
